cdptools/bin/clone_file_store.py

```diff
@@ -0,0 +1,73 @@
+import argparse
+import logging
+from pathlib import Path
+
+from google.api_core.page_iterator import Page
+
+from concurrent.futures import ThreadPoolExecutor
+from functools import partial
+import cdptools.file_stores.gcs_file_store as fs
+
+logging.basicConfig(
+    level=logging.INFO,
+    format="[%(levelname)4s: %(module)s:%(lineno)4s %(asctime)s] %(message)s"
+)
+log = logging.getLogger(__name__)
+
+###############################################################################
+
+
+class Args(argparse.Namespace):
+
+    def __init__(self):
+        self.__parse()
+
+    def __parse(self):
+        p = argparse.ArgumentParser(
+            prog="clone_file_store",
+            description="Clone source file store to target file store."
+        )
+
+        p.add_argument("--target_bucket", type=str, help="Target bucket name.", required=True)
+        p.add_argument("--target_credentials", type=Path, help="Credentials of target file store.", required=True)
+        p.add_argument("--debug", action="store_true", dest="debug", help="Show traceback if the script were to fail.")
+        p.add_argument('--source_bucket', help="Source bucket name.", required=True)
+        p.add_argument('--source_credentials', type=Path, help="Credentials of source file store.")
+
+        p.parse_args(namespace=self)
+
+
+def clone_page(page: Page, source_fs: fs.GCSFileStore, target_fs: fs.GCSFileStore):
+    for blob in page:
+        # download file takes in a filename and returns a path
+        downloaded_path = source_fs.download_file(blob.name, overwrite=True)
+
+        # upload file takes in a path
+        target_fs.upload_file(downloaded_path)
+
+
+def main():
+    args = Args()
+
+    source_fs = fs.GCSFileStore(args.source_bucket, args.source_credentials)
+    target_fs = fs.GCSFileStore(args.target_bucket, args.target_credentials)
+
+    # We call client list_blobs method as use of the bucket's list_blobs method is deprecated
+    source_client = source_fs._client
+    source_bucket = source_fs._bucket
+
+    # Clear the target file store of existing files
+    target_fs.clear_bucket()
+
+    # Get blobs in pages so we can use threading
+    pages = source_client.list_blobs(source_bucket).pages
```

Comment:
Adding to above comment, I think the solution may just be to add a function to each database that is called like "file iterator". That way it can return files however it wants but as long as we do end up getting a file we can multithread delete it.

Reply:
For this case with `CloudFirestoreDatabase` I'm currently multithreading on pages (which each contain files), but do you think it'd be better to multithread on individual files themselves? I think the multithreading on individual files will most likely be faster in the case of a small number of `pages`, and I think a `page` can get fairly large. 

The reason why I threaded on pages originally was because I was concerned that loading all the files into one list could get too large. Do you think this would be an issue?

If we thread on individual files, creating an abstract db function that returns a db's entire storage would be a good solution. Also by file iterator do you mean just a function that returns a list of files, or an actual iterator object?